File: Example/ios-example-objective-c/ViewController.m
//
//  ViewController.m
//  ios-sdk-example
//
//  Created by Christopher Kleeschulte on 4/10/15.
//  Copyright (c) 2015 Christopher Kleeschulte. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()
@property (readonly) NSString *key;
@property (readonly) NSString *sin;
@property (readonly) NSString *token;
@property (readonly) NSString *signedMessage;
@property (readonly) NSString *invoice;
@property (nonatomic, strong) NSMutableData *responseData;
@end

@implementation ViewController

NSString *genericErrorMessage = @"There was an error from the api";
NSString *bitpayUrl = @"https://test.bitpay.com";

- (void)viewDidLoad {
    [super viewDidLoad];
    _pairText.delegate = self;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

- (IBAction)generateKeys:(id)sender {
    _key = [BPKeyUtils generatePem];
    _keyText.text = _key;
    NSLog(@"pem key: %@", _key);
}

- (IBAction)generateSin:(id)sender {
    if(_key == nil) {
        _sinText.text = @"Please generate a key in the previous step first.";
    } else {
        _sin = [BPKeyUtils generateSinFromPem:_key];
        _sinText.text = _sin;
        NSLog(@"sin: %@", _sin);
    }
}

- (IBAction)getToken:(id)sender {

    if(_pairText.text == nil) {
        _tokenText.text = @"Please get a pairing code from test.bitpay.com";
    } else {
        
        NSLog(@"pairingCode: %@", _pairText.text);
        
        BPBitPay *bp = [[BPBitPay alloc] initWithName:@"BitPay" pem: _key];
        bp.host = bitpayUrl;
        
        NSError *error = nil;
        NSString *token = [bp authorizeClient:_pairText.text error: &error];
        
        if(error) {
            NSLog(@"%@", @"there was an error");
            _tokenText.text = @"there was an error";
            return;
        }
        
        _token = token;
        _tokenText.text = token;
        
    }
    
}

- (IBAction)testCreatInvoice:(id)sender {
    
    if(_token == nil || _key == nil || _sin == nil) {
        _invoiceText.text = @"Please geneate a key/sin/token before creating an invoice here.";
        return;
    }
    
    NSString *pubKey = [BPKeyUtils getPublicKeyFromPem:_key];
    NSLog(@"public key: %@", pubKey);
    NSLog(@"private key: %@", [BPKeyUtils getPrivateKeyFromPem:_key]);
    
    NSString *postString = [NSString stringWithFormat:@"{\"currency\":\"USD\",\"price\":20,\"token\":\"%@\"}", _token];
    
    NSString *message = [NSString stringWithFormat: @"%@/invoices%@", bitpayUrl, postString];
    
    _signedMessage = [BPKeyUtils sign:message withPem:_key];
    
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL: [NSURL URLWithString:[NSString stringWithFormat:@"%@/invoices", bitpayUrl]]];
    
    [request setHTTPBody:[postString dataUsingEncoding:NSUTF8StringEncoding]];
    [request addValue:@"2.0.0" forHTTPHeaderField:@"x-accept-version"];
    [request addValue:@"application/json" forHTTPHeaderField:@"content-type"];
    [request addValue:pubKey forHTTPHeaderField:@"x-identity"];
    [request addValue:_signedMessage forHTTPHeaderField:@"x-signature"];
    [request setHTTPMethod:@"POST"];
    
    [NSURLConnection connectionWithRequest:request delegate:self];

}

#pragma mark UITextFieldDelegate

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    [textField resignFirstResponder];
    return NO;
}


#pragma mark NSURLConnection Delegate Methods

- (void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLResponse *)response {
    _responseData = [[NSMutableData alloc] init];
}

- (void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data {
    [_responseData appendData:data];
}

- (NSCachedURLResponse *)connection:(NSURLConnection *)connection
                  willCacheResponse:(NSCachedURLResponse*)cachedResponse {
    return nil;
}

- (void)connectionDidFinishLoading:(NSURLConnection *)connection {
    
    NSError *error = nil;
    
    id object = [NSJSONSerialization
                 JSONObjectWithData:_responseData
                 options:0
                 error:&error];
    
    if(error) {
        _tokenText.text = genericErrorMessage;
    }
    
    if([object isKindOfClass:[NSDictionary class]]) {
        
        NSDictionary *results = object;
        NSString *error = [results objectForKey:@"error"];
        
        if(error) {
            NSLog(@"error: %@", error);
            _invoiceText.text = error;
            return;
        }
        
        NSDictionary *data = [results valueForKeyPath: @"data"];
        
        if(data) {
            id tokenId = [data valueForKeyPath: @"token"];
            if([tokenId isKindOfClass:[NSArray class]]) {
                NSArray *tokenArray = tokenId;
                _token = tokenArray[0];
                _tokenText.text = _token;
            }
            _invoiceText.text = [results description];
        } else {
            _invoiceText.text = [results description];
            NSLog(@"%@", [results description]);
            return;
        }
    }

}

- (void)connection:(NSURLConnection *)connection didFailWithError:(NSError *)error {
    NSLog(@"there was an error.");
}



@end

File: Example/ios-example-swift/ViewController.swift
//
//  ViewController.swift
//  ios-example-swift
//
//  Created by chrisk on 4/15/15.
//  Copyright (c) 2015 bitpay. All rights reserved.
//

import UIKit

class ViewController: UIViewController,  NSURLConnectionDelegate, UITextFieldDelegate {

    @IBOutlet weak var pairingText: UITextField!
    @IBOutlet weak var pemText: UITextView!
    @IBOutlet weak var sinText: UITextField!
    @IBOutlet weak var tokenText: UITextField!
    @IBOutlet weak var invoiceText: UITextView!
    
    var pem = ""
    var key = ""
    var sin = ""
    var token = ""
    var invoice = ""
    var data = NSMutableData()
    
    @IBAction func generateKey(sender: AnyObject) {
        pem = BPKeyUtils.generatePem()
        pemText.text = pem
    }
    
    
    @IBAction func generateSin(sender: AnyObject) {
        sin = BPKeyUtils.generateSinFromPem(pem)
        sinText.text = sin
    }
    
    @IBAction func getToken(sender: AnyObject) {
        
        let pairingCode = pairingText.text as String!
        if(pairingCode.isEmpty) {
            tokenText.text = "Please get a pairing code from test.bitpay.com"
            return
        }

        data.setData(NSData())

        NSLog("pairingCode: \(pairingText.text)")
        
        let bp = BPBitPay(name: "BitPay", pem: pem)
        
        bp.host = "https://test.bitpay.com"
        
        do {
            let token = try bp.authorizeClient(pairingCode)
            self.token = token
            tokenText.text = token

        } catch _ {
            NSLog("api error")
            tokenText.text = "api error"
        }
    }
    
    
    @IBAction func testCreateInvoice(sender: AnyObject) {
        
        if(token.isEmpty || pem.isEmpty) {
            tokenText.text = "Please uses steps 2 and 4 before attempting to test the invoice"
            return
        }
        
        data.setData(NSData())
        
        let url = NSURL(string: "https://test.bitpay.com/invoices")
        let request = NSMutableURLRequest(URL: url!)

        
        let pubkey = BPKeyUtils.getPublicKeyFromPem(pem)
        NSLog("public key: \(pubkey)")
        NSLog("private key: \(BPKeyUtils.getPrivateKeyFromPem(pem))")
        
        let postString = "{\"currency\":\"USD\",\"price\":20,\"token\":\"\(token)\"}"
        
        let message = "https://test.bitpay.com/invoices\(postString)"
        
        let signedMessage = BPKeyUtils.sign(message, withPem: pem)

        let bodyData = (postString as NSString).dataUsingEncoding(NSUTF8StringEncoding)
        
        request.HTTPBody = bodyData
        request.addValue("2.0.0", forHTTPHeaderField: "x-accept-version")
        request.addValue(pubkey, forHTTPHeaderField: "x-identity")
        request.addValue("application/json", forHTTPHeaderField: "content-type")
        request.addValue(signedMessage, forHTTPHeaderField: "x-signature")
        request.HTTPMethod = "POST"
        
        let connection = NSURLConnection(request: request, delegate: self, startImmediately: true)
        connection?.start()
        
    }
    
//UITextFieldDelegate Methods
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        self.view.endEditing(true)
        return false
    }
//end

    
    override func viewDidLoad() {
        super.viewDidLoad()
        let tap: UITapGestureRecognizer = UITapGestureRecognizer(target: self, action: "dismissKeyboard")
        view.addGestureRecognizer(tap)
        self.pairingText.delegate = self
    }

    func dismissKeyboard() {
        view.endEditing(true)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        
    }
    
//NSURLConnection Delegate Methods
    func connection(connection: NSURLConnection,
        didReceiveData data: NSData!) {
            self.data.appendData(data)
    }
    
    
    func connectionDidFinishLoading(connection: NSURLConnection!) {

        var jsonError: NSError?
        let errorInParsing = "there was an error parsing"
        
        let obj: AnyObject?
        do {
            obj = try NSJSONSerialization.JSONObjectWithData(data,
                        options: NSJSONReadingOptions.AllowFragments)
        } catch let error as NSError {
            jsonError = error
            obj = nil
        }

        let dta = NSString(data: data, encoding: NSUTF8StringEncoding)
        print("response from server is: \(dta)")
        
        if(jsonError != nil) {
            NSLog(errorInParsing)
            invoiceText.text = errorInParsing
            return
        }
        
        if let root = obj as? NSDictionary {
            if let data = root["data"] as? NSArray {
                if let tokenDictionary = data[0] as? NSDictionary {
                    if let token = tokenDictionary["token"] as? NSString {
                        self.token = token as String
                        tokenText.text = self.token
                    }
                }
            } else {
                //this must be the invoice creation call
                invoiceText.text = root.description as String
            }
            
        }
        
    }
    
    func connection(connection: NSURLConnection, didFinishWithError error: NSErrorPointer) {
        invoiceText.text = "There was an error from the api."
    }

    

}